File: shuvoice/waybar/format.py
"""Formatting helpers for Waybar JSON + tooltip content."""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any

from ..config import DEFAULT_ELEVENLABS_TTS_VOICE_ID, Config
from ..tts_speed import format_tts_playback_speed

_OPENAI_VOICE_LABELS = {
    "alloy": "Alloy",
    "ash": "Ash",
    "coral": "Coral",
    "echo": "Echo",
    "fable": "Fable",
    "nova": "Nova",
    "onyx": "Onyx",
    "sage": "Sage",
    "shimmer": "Shimmer",
}


def _tts_backend_label(backend: str) -> str:
    return {
        "elevenlabs": "ElevenLabs",
        "openai": "OpenAI",
        "local": "Local Piper",
    }.get(str(backend).strip().lower(), str(backend).strip() or "unknown")


def _tts_voice_label(backend: str, voice_id: str) -> str:
    backend_id = str(backend).strip().lower()
    voice = str(voice_id).strip()
    if backend_id == "openai":
        return _OPENAI_VOICE_LABELS.get(voice.lower(), voice or "default")
    if backend_id == "elevenlabs" and voice == DEFAULT_ELEVENLABS_TTS_VOICE_ID:
        return "Default"
    if backend_id == "local" and not voice:
        return "configured local default"
    return voice or "default"


def config_info_lines(config: Config) -> list[str]:
    """Build tooltip info lines from the active configuration."""
    backend = config.asr_backend
    backend_labels = {
        "nemo": "NeMo (NVIDIA)",
        "sherpa": "Sherpa-ONNX",
        "moonshine": "Moonshine-ONNX",
    }

    lines = [f"Backend:  {backend_labels.get(backend, backend)}"]
    if config.instant_mode:
        lines.append("Profile:  Instant")

    # Model name (shortened for readability)
    if backend == "nemo":
        model = config.model_name
        if "/" in model:
            model = model.rsplit("/", 1)[-1]
        lines.append(f"Model:    {model}")
    elif backend == "sherpa":
        if config.sherpa_model_dir:
            model = Path(config.sherpa_model_dir).name
        else:
            model = f"{config.sherpa_model_name} (auto-download)"
        lines.append(f"Model:    {model}")
    elif backend == "moonshine":
        lines.append(f"Model:    {config.moonshine_model_name}")

    # Device (GPU vs CPU)
    if backend == "nemo":
        device = "GPU (CUDA)" if config.device == "cuda" else "CPU"
    elif backend == "sherpa":
        device = "GPU (CUDA)" if config.sherpa_provider == "cuda" else "CPU"
    elif backend == "moonshine":
        device = "GPU (CUDA)" if config.moonshine_provider == "cuda" else "CPU"
    else:
        device = "unknown"
    lines.append(f"Device:   {device}")
    if config.tts_enabled:
        lines.append(f"TTS:      {_tts_backend_label(config.tts_backend)}")
        lines.append(f"Voice:    {_tts_voice_label(config.tts_backend, config.tts_default_voice_id)}")
        lines.append(f"Speed:    {format_tts_playback_speed(config.tts_playback_speed)}")
    else:
        lines.append("TTS:      Disabled")

    return lines


def sanitize_class(value: str) -> str:
    cleaned = re.sub(r"[^a-z0-9_-]+", "-", value.lower()).strip("-")
    return cleaned or "unknown"


def build_waybar_payload(
    state: str,
    *,
    config_lines: list[str] | None = None,
    service_state: str | None = None,
    control_error: str | None = None,
    action_error: str | None = None,
) -> dict[str, Any]:
    base_state = state
    reason = ""
    if ":" in state:
        base_state, reason = state.split(":", 1)

    icons = {
        "recording": "",
        "processing": "",
        "idle": "",
        "starting": "",
        "stopped": "",
        "error": "",
    }

    labels = {
        "recording": "Recording",
        "processing": "Processing",
        "idle": "Ready",
        "starting": "Starting",
        "stopped": "Stopped",
        "error": "Error",
    }

    if base_state not in labels:
        base_state = "error"
        reason = reason or "unknown_state"

    lines = [f"ShuVoice: {labels[base_state]}"]

    if reason:
        lines.append(f"Reason: {reason}")
    if service_state and service_state != "unknown":
        lines.append(f"Service: {service_state}")
    if control_error and base_state in {"starting", "error"}:
        lines.append(f"Control: {control_error}")
    if action_error:
        lines.append(f"Action: {action_error}")

    if config_lines:
        lines.append("")
        lines.extend(config_lines)

    lines.extend(
        [
            "",
            "Left click: toggle recording",
            "Middle click: toggle service",
            "Right click: open action menu",
        ]
    )

    class_name = sanitize_class(base_state)

    return {
        "text": icons[base_state],
        "alt": base_state,
        "class": class_name,
        "tooltip": "\n".join(lines),
    }

File: crates/shuvoice-io/src/waybar/format.rs
//! Waybar JSON + tooltip formatting.

use std::collections::BTreeMap;

use regex::Regex;
use serde_json::{Value, json};
use std::sync::OnceLock;

fn class_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[^a-z0-9_-]+").expect("static"))
}

/// Sanitize a CSS-ish class token.
#[must_use]
pub fn sanitize_class(value: &str) -> String {
    let cleaned = class_re()
        .replace_all(&value.to_ascii_lowercase(), "-")
        .trim_matches('-')
        .to_string();
    if cleaned.is_empty() {
        "unknown".into()
    } else {
        cleaned
    }
}

/// Minimal config projection for tooltip lines (avoids full core Config coupling).
#[derive(Debug, Clone, Default)]
pub struct WaybarConfigInfo {
    pub asr_backend: String,
    pub instant_mode: bool,
    pub model_label: Option<String>,
    pub device_label: Option<String>,
    pub tts_enabled: bool,
    pub tts_backend_label: Option<String>,
    pub tts_voice_label: Option<String>,
    pub tts_speed_label: Option<String>,
    pub overlay_debug_mode: bool,
}

/// Build tooltip info lines from config projection.
#[must_use]
pub fn config_info_lines(cfg: &WaybarConfigInfo) -> Vec<String> {
    let backend_label = match cfg.asr_backend.as_str() {
        "nemo" => "NeMo (NVIDIA)".into(),
        "sherpa" => "Sherpa-ONNX".into(),
        "moonshine" => "Moonshine-ONNX".into(),
        other => other.to_string(),
    };
    let mut lines = vec![format!("Backend:  {backend_label}")];
    if cfg.instant_mode {
        lines.push("Profile:  Instant".into());
    }
    if let Some(model) = &cfg.model_label {
        lines.push(format!("Model:    {model}"));
    }
    if let Some(device) = &cfg.device_label {
        lines.push(format!("Device:   {device}"));
    }
    if cfg.tts_enabled {
        lines.push(format!(
            "TTS:      {}",
            cfg.tts_backend_label.as_deref().unwrap_or("unknown")
        ));
        if let Some(voice) = &cfg.tts_voice_label {
            lines.push(format!("Voice:    {voice}"));
        }
        if let Some(speed) = &cfg.tts_speed_label {
            lines.push(format!("Speed:    {speed} (default synth)"));
        }
    } else {
        lines.push("TTS:      Disabled".into());
    }
    lines.push(format!(
        "Debug:    {}",
        if cfg.overlay_debug_mode {
            "Overlay on"
        } else {
            "Overlay off"
        }
    ));
    lines
}

/// Build the Waybar custom-module JSON object.
#[must_use]
pub fn build_waybar_payload(
    state: &str,
    config_lines: Option<&[String]>,
    service_state: Option<&str>,
    control_error: Option<&str>,
    action_error: Option<&str>,
) -> Value {
    let (mut base_state, mut reason) = match state.split_once(':') {
        Some((b, r)) => (b.to_string(), r.to_string()),
        None => (state.to_string(), String::new()),
    };

    let mut icons = BTreeMap::new();
    icons.insert("recording", "\u{f130}"); // 
    icons.insert("processing", "\u{f252}"); // 
    icons.insert("idle", "\u{f130}");
    icons.insert("starting", "\u{f252}");
    icons.insert("stopped", "\u{f131}"); // 
    icons.insert("error", "\u{f071}"); // 

    let mut labels = BTreeMap::new();
    labels.insert("recording", "Recording");
    labels.insert("processing", "Processing");
    labels.insert("idle", "Ready");
    labels.insert("starting", "Starting");
    labels.insert("stopped", "Stopped");
    labels.insert("error", "Error");

    if !labels.contains_key(base_state.as_str()) {
        if reason.is_empty() {
            reason = "unknown_state".into();
        }
        base_state = "error".into();
    }

    let mut lines = vec![format!(
        "ShuVoice: {}",
        labels.get(base_state.as_str()).copied().unwrap_or("Error")
    )];
    if !reason.is_empty() {
        lines.push(format!("Reason: {reason}"));
    }
    if let Some(ss) = service_state
        && ss != "unknown"
    {
        lines.push(format!("Service: {ss}"));
    }
    if let Some(ce) = control_error
        && (base_state == "starting" || base_state == "error")
    {
        lines.push(format!("Control: {ce}"));
    }
    if let Some(ae) = action_error {
        lines.push(format!("Action: {ae}"));
    }
    if let Some(cfg_lines) = config_lines {
        lines.push(String::new());
        lines.extend(cfg_lines.iter().cloned());
    }
    lines.push(String::new());
    lines.push("Left click: toggle recording".into());
    lines.push("Middle click: toggle service".into());
    lines.push("Right click: open action menu".into());

    let class_name = sanitize_class(&base_state);
    let text = icons
        .get(base_state.as_str())
        .copied()
        .unwrap_or("\u{f071}");

    json!({
        "text": text,
        "alt": base_state,
        "class": class_name,
        "tooltip": lines.join("\n"),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn recording_payload() {
        let p = build_waybar_payload("recording", None, None, None, None);
        assert_eq!(p["alt"], "recording");
        assert_eq!(p["class"], "recording");
        assert!(
            p["tooltip"]
                .as_str()
                .unwrap()
                .contains("ShuVoice: Recording")
        );
    }

    #[test]
    fn error_reason_sanitized_class() {
        let p = build_waybar_payload(
            "error:asr_thread_dead",
            None,
            Some("failed"),
            Some("Control socket not found"),
            Some("systemctl restart failed"),
        );
        assert_eq!(p["alt"], "error");
        assert_eq!(p["class"], "error");
        let tip = p["tooltip"].as_str().unwrap();
        assert!(tip.contains("Reason: asr_thread_dead"));
        assert!(tip.contains("Service: failed"));
        assert!(tip.contains("Control: Control socket not found"));
    }

    #[test]
    fn config_lines_include_debug() {
        let cfg = WaybarConfigInfo {
            asr_backend: "sherpa".into(),
            overlay_debug_mode: true,
            tts_enabled: false,
            ..Default::default()
        };
        let lines = config_info_lines(&cfg);
        assert!(lines.iter().any(|l| l.contains("Sherpa-ONNX")));
        assert!(lines.iter().any(|l| l.contains("Overlay on")));
        assert!(lines.iter().any(|l| l.contains("TTS:      Disabled")));
    }
}
